File: DRONE/maps/generate_urban_map.py
#!/usr/bin/env python3
"""
Urban City Map Generator
Creates a realistic urban environment map for drone simulation
"""

import numpy as np
import cv2
import random
from PIL import Image

class UrbanMapGenerator:
    def __init__(self, width=2000, height=2000, resolution=0.05):
        """
        Initialize urban map generator
        Args:
            width: Map width in pixels
            height: Map height in pixels  
            resolution: Meters per pixel (0.05 = 5cm per pixel)
        """
        self.width = width
        self.height = height
        self.resolution = resolution
        self.map_data = np.full((height, width), 255, dtype=np.uint8)  # Start with free space
        
        # Color codes for different areas
        self.OCCUPIED = 0      # Black - buildings/obstacles
        self.FREE = 255        # White - free space
        self.UNKNOWN = 128     # Gray - unknown areas
        
    def add_building_block(self, x, y, width, height, building_type='normal'):
        """Add a building block to the map"""
        x1, y1 = max(0, x), max(0, y)
        x2, y2 = min(self.width, x + width), min(self.height, y + height)
        
        if building_type == 'highrise':
            # High-rise buildings (completely solid)
            self.map_data[y1:y2, x1:x2] = self.OCCUPIED
        else:
            # Regular buildings with some internal structure
            self.map_data[y1:y2, x1:x2] = self.OCCUPIED
            # Add some internal courtyards for larger buildings
            if width > 100 and height > 100:
                courtyard_x = x1 + width//3
                courtyard_y = y1 + height//3
                courtyard_w = width//3
                courtyard_h = height//3
                self.map_data[courtyard_y:courtyard_y+courtyard_h, 
                             courtyard_x:courtyard_x+courtyard_w] = self.FREE
    
    def add_road(self, x1, y1, x2, y2, width=40):
        """Add a road between two points"""
        # Calculate road direction and create perpendicular width
        dx = x2 - x1
        dy = y2 - y1
        length = np.sqrt(dx*dx + dy*dy)
        
        if length == 0:
            return
            
        # Normalize direction
        dx_norm = dx / length
        dy_norm = dy / length
        
        # Perpendicular direction for road width
        perp_x = -dy_norm * width // 2
        perp_y = dx_norm * width // 2
        
        # Create road polygon
        points = np.array([
            [x1 + perp_x, y1 + perp_y],
            [x1 - perp_x, y1 - perp_y],
            [x2 - perp_x, y2 - perp_y],
            [x2 + perp_x, y2 + perp_y]
        ], dtype=np.int32)
        
        cv2.fillPoly(self.map_data, [points], self.FREE)
    
    def add_intersection(self, x, y, size=60):
        """Add a road intersection"""
        x1, y1 = max(0, x - size//2), max(0, y - size//2)
        x2, y2 = min(self.width, x + size//2), min(self.height, y + size//2)
        self.map_data[y1:y2, x1:x2] = self.FREE
    
    def add_park(self, x, y, width, height):
        """Add a park area (partially free space with some trees)"""
        x1, y1 = max(0, x), max(0, y)
        x2, y2 = min(self.width, x + width), min(self.height, y + height)
        
        # Park is mostly free space
        self.map_data[y1:y2, x1:x2] = self.FREE
        
        # Add some trees/obstacles randomly in the park
        for _ in range((width * height) // 5000):  # Tree density
            tree_x = random.randint(x1 + 10, x2 - 10)
            tree_y = random.randint(y1 + 10, y2 - 10)
            tree_size = random.randint(8, 15)
            cv2.circle(self.map_data, (tree_x, tree_y), tree_size, self.OCCUPIED, -1)
    
    def add_parking_lot(self, x, y, width, height):
        """Add a parking lot with some cars"""
        x1, y1 = max(0, x), max(0, y)
        x2, y2 = min(self.width, x + width), min(self.height, y + height)
        
        # Parking lot is free space
        self.map_data[y1:y2, x1:x2] = self.FREE
        
        # Add parked cars
        car_width, car_height = 8, 4  # Car size in pixels
        for car_y in range(y1 + 10, y2 - 10, 20):
            for car_x in range(x1 + 10, x2 - 10, 15):
                if random.random() < 0.6:  # 60% occupancy
                    self.map_data[car_y:car_y+car_height, car_x:car_x+car_width] = self.OCCUPIED
    
    def add_landing_zone(self, x, y, size=30):
        """Add a designated landing zone"""
        x1, y1 = max(0, x - size//2), max(0, y - size//2)
        x2, y2 = min(self.width, x + size//2), min(self.height, y + size//2)
        
        # Clear the area
        self.map_data[y1:y2, x1:x2] = self.FREE
        
        # Add landing pad border
        cv2.rectangle(self.map_data, (x1, y1), (x2, y2), self.OCCUPIED, 2)
        
        # Add cross marker in center
        cv2.line(self.map_data, (x-10, y), (x+10, y), self.OCCUPIED, 2)
        cv2.line(self.map_data, (x, y-10), (x, y+10), self.OCCUPIED, 2)
    
    def generate_urban_layout(self):
        """Generate a realistic urban layout"""
        print("Generating urban city layout...")
        
        # 1. Create main road grid
        print("Adding road network...")
        
        # Vertical main roads
        for x in [400, 800, 1200, 1600]:
            self.add_road(x, 0, x, self.height, 50)  # Main roads
        
        # Horizontal main roads  
        for y in [400, 800, 1200, 1600]:
            self.add_road(0, y, self.width, y, 50)
            
        # Secondary roads
        for x in [200, 600, 1000, 1400, 1800]:
            self.add_road(x, 0, x, self.height, 30)
            
        for y in [200, 600, 1000, 1400, 1800]:
            self.add_road(0, y, self.width, y, 30)
        
        # Add intersections
        print("Adding intersections...")
        road_positions = [200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800]
        for x in road_positions:
            for y in road_positions:
                self.add_intersection(x, y, 80)
        
        # 2. Add downtown high-rise district
        print("Adding downtown district...")
        for x in range(300, 700, 80):
            for y in range(300, 700, 80):
                if random.random() < 0.8:  # 80% building density
                    building_w = random.randint(60, 120)
                    building_h = random.randint(60, 120)
                    self.add_building_block(x, y, building_w, building_h, 'highrise')
        
        # 3. Add residential areas
        print("Adding residential areas...")
        # Top-left residential
        for x in range(50, 350, 40):
            for y in range(50, 350, 40):
                if random.random() < 0.6:
                    building_w = random.randint(25, 50)
                    building_h = random.randint(25, 50)
                    self.add_building_block(x, y, building_w, building_h)
        
        # Top-right residential
        for x in range(1650, 1950, 40):
            for y in range(50, 350, 40):
                if random.random() < 0.6:
                    building_w = random.randint(25, 50)
                    building_h = random.randint(25, 50)
                    self.add_building_block(x, y, building_w, building_h)
        
        # Bottom areas
        for x in range(50, 1950, 60):
            for y in range(1650, 1950, 60):
                if random.random() < 0.4:
                    building_w = random.randint(30, 80)
                    building_h = random.randint(30, 80)
                    self.add_building_block(x, y, building_w, building_h)
        
        # 4. Add commercial/industrial areas
        print("Adding commercial areas...")
        for x in range(900, 1500, 100):
            for y in range(50, 300, 100):
                if random.random() < 0.7:
                    building_w = random.randint(80, 150)
                    building_h = random.randint(60, 100)
                    self.add_building_block(x, y, building_w, building_h)
        
        # 5. Add parks and green spaces
        print("Adding parks...")
        # Central park
        self.add_park(850, 850, 300, 300)
        
        # Smaller parks
        self.add_park(50, 450, 120, 120)
        self.add_park(1650, 450, 120, 120)
        self.add_park(450, 1650, 120, 120)
        
        # 6. Add parking lots
        print("Adding parking areas...")
        self.add_parking_lot(750, 300, 100, 80)
        self.add_parking_lot(300, 750, 80, 100)
        self.add_parking_lot(1300, 1300, 120, 100)
        self.add_parking_lot(1500, 500, 100, 80)
        
        # 7. Add landing zones
        print("Adding landing zones...")
        # Hospital landing pad
        self.add_landing_zone(1000, 1000, 40)
        
        # Emergency landing zones
        self.add_landing_zone(100, 100, 30)
        self.add_landing_zone(1900, 100, 30)
        self.add_landing_zone(100, 1900, 30)
        self.add_landing_zone(1900, 1900, 30)
        
        # Police station landing
        self.add_landing_zone(500, 500, 35)
        
        print("Urban layout generation completed!")
    
    def save_pgm(self, filename):
        """Save map as PGM file"""
        print(f"Saving map as {filename}...")
        
        # Create PIL image and save as PGM
        img = Image.fromarray(self.map_data, mode='L')
        img.save(filename)
        
        print(f"Map saved successfully: {filename}")
        print(f"Map size: {self.width}x{self.height} pixels")
        print(f"Real world size: {self.width*self.resolution:.1f}m x {self.height*self.resolution:.1f}m")
        print(f"Resolution: {self.resolution*100:.1f}cm per pixel")

def main():
    """Generate urban city map"""
    print("=== Urban City Map Generator ===")
    
    # Create map generator
    generator = UrbanMapGenerator(width=2000, height=2000, resolution=0.05)
    
    # Generate urban layout
    generator.generate_urban_layout()
    
    # Save map
    generator.save_pgm("urban_city.pgm")
    
    print("\n=== Generation Complete ===")
    print("Files created:")
    print("- urban_city.pgm (map data)")
    print("- urban_city.yaml (map configuration)")
    print("- urban_city_preview.txt (map description)")
    print("\nTo use this map:")
    print("1. Copy files to your ROS2 maps directory")
    print("2. Update your launch file to use 'urban_city.yaml'")
    print("3. Launch your navigation stack")

if __name__ == "__main__":
    main() 

File: DRONE/maps/generate_urban_map.m
function generate_urban_map()
% Urban City Map Generator for Drone Navigation
% Creates a realistic urban environment using MATLAB
% Output: PGM format map file for ROS navigation

    % Map parameters
    map_width = 2000;   % pixels (100m at 5cm resolution)
    map_height = 2000;  % pixels
    resolution = 0.05;  % meters per pixel
    
    % Initialize map with free space (white = 255)
    city_map = 255 * ones(map_height, map_width, 'uint8');
    
    fprintf('Generating Urban City Map...\n');
    fprintf('Map size: %dx%d pixels (%.1fm x %.1fm)\n', ...
            map_width, map_height, map_width*resolution, map_height*resolution);
    
    % Add main roads (horizontal and vertical grid)
    city_map = add_road_network(city_map, map_width, map_height);
    
    % Add downtown high-rise district (center)
    city_map = add_downtown_district(city_map, map_width, map_height);
    
    % Add residential areas (corners)
    city_map = add_residential_areas(city_map, map_width, map_height);
    
    % Add parks and green spaces
    city_map = add_parks(city_map, map_width, map_height);
    
    % Add parking lots
    city_map = add_parking_areas(city_map, map_width, map_height);
    
    % Add emergency landing zones
    city_map = add_landing_zones(city_map, map_width, map_height);
    
    % Save as PGM file
    save_pgm_file(city_map, 'urban_city_improved.pgm');
    
    % Create visualization
    create_visualization(city_map);
    
    fprintf('Urban city map generated successfully!\n');
    fprintf('Files created:\n');
    fprintf('- urban_city_improved.pgm (map data)\n');
    fprintf('- urban_city_visualization.png (preview)\n');
end

function map = add_road_network(map, width, height)
    % Add main roads in a grid pattern
    road_width = 40;  % pixels (2m wide roads)
    
    % Vertical roads every 400 pixels (20m apart)
    for x = 200:400:width-200
        x_start = max(1, x - road_width/2);
        x_end = min(width, x + road_width/2);
        map(:, x_start:x_end) = 128;  % Gray for roads
    end
    
    % Horizontal roads every 400 pixels (20m apart)
    for y = 200:400:height-200
        y_start = max(1, y - road_width/2);
        y_end = min(height, y + road_width/2);
        map(y_start:y_end, :) = 128;  % Gray for roads
    end
    
    % Add intersections (darker gray)
    for x = 200:400:width-200
        for y = 200:400:height-200
            x_start = max(1, x - road_width/2);
            x_end = min(width, x + road_width/2);
            y_start = max(1, y - road_width/2);
            y_end = min(height, y + road_width/2);
            map(y_start:y_end, x_start:x_end) = 100;  % Darker gray for intersections
        end
    end
end

function map = add_downtown_district(map, width, height)
    % Add high-rise buildings in the center area
    center_x = width/2;
    center_y = height/2;
    district_size = 600;  % 30m radius
    
    % Create several large buildings
    buildings = [
        center_x-200, center_y-200, 120, 120;  % Building 1
        center_x-50,  center_y-180, 100, 140;  % Building 2
        center_x+100, center_y-150, 80,  100;  % Building 3
        center_x-180, center_y+50,  140, 100;  % Building 4
        center_x+50,  center_y+80,  120, 120;  % Building 5
    ];
    
    for i = 1:size(buildings, 1)
        x = round(buildings(i, 1));
        y = round(buildings(i, 2));
        w = buildings(i, 3);
        h = buildings(i, 4);
        
        x_start = max(1, x);
        x_end = min(width, x + w);
        y_start = max(1, y);
        y_end = min(height, y + h);
        
        map(y_start:y_end, x_start:x_end) = 0;  % Black for buildings
    end
end

function map = add_residential_areas(map, width, height)
    % Add smaller residential buildings in corners
    building_size = 60;  % 3m x 3m buildings
    spacing = 120;       % 6m spacing
    
    % Top-left residential area
    for x = 50:spacing:400
        for y = 50:spacing:400
            if map(y, x) == 255  % Only place on free space
                x_end = min(width, x + building_size);
                y_end = min(height, y + building_size);
                map(y:y_end, x:x_end) = 50;  % Dark gray for residential
            end
        end
    end
    
    % Top-right residential area
    for x = width-400:spacing:width-50
        for y = 50:spacing:400
            if x <= width && y <= height && map(y, x) == 255
                x_end = min(width, x + building_size);
                y_end = min(height, y + building_size);
                map(y:y_end, x:x_end) = 50;
            end
        end
    end
    
    % Bottom areas (similar pattern)
    for x = 50:spacing:400
        for y = height-400:spacing:height-50
            if y <= height && map(y, x) == 255
                x_end = min(width, x + building_size);
                y_end = min(height, y + building_size);
                map(y:y_end, x:x_end) = 50;
            end
        end
    end
    
    for x = width-400:spacing:width-50
        for y = height-400:spacing:height-50
            if x <= width && y <= height && map(y, x) == 255
                x_end = min(width, x + building_size);
                y_end = min(height, y + building_size);
                map(y:y_end, x:x_end) = 50;
            end
        end
    end
end

function map = add_parks(map, width, height)
    % Add park areas (lighter gray for navigation)
    parks = [
        width*0.25, height*0.75, 200, 150;  % Park 1
        width*0.75, height*0.25, 180, 160;  % Park 2
        width*0.15, height*0.15, 120, 100;  % Small park 3
        width*0.85, height*0.85, 140, 120;  % Small park 4
    ];
    
    for i = 1:size(parks, 1)
        x = round(parks(i, 1));
        y = round(parks(i, 2));
        w = parks(i, 3);
        h = parks(i, 4);
        
        x_start = max(1, x);
        x_end = min(width, x + w);
        y_start = max(1, y);
        y_end = min(height, y + h);
        
        % Only place parks on free space
        for py = y_start:y_end
            for px = x_start:x_end
                if map(py, px) == 255
                    map(py, px) = 200;  % Light gray for parks
                end
            end
        end
    end
end

function map = add_parking_areas(map, width, height)
    % Add parking lots near buildings
    parking_areas = [
        width*0.3, height*0.4, 100, 80;
        width*0.7, height*0.6, 120, 90;
        width*0.2, height*0.8, 80, 70;
        width*0.8, height*0.2, 90, 85;
    ];
    
    for i = 1:size(parking_areas, 1)
        x = round(parking_areas(i, 1));
        y = round(parking_areas(i, 2));
        w = parking_areas(i, 3);
        h = parking_areas(i, 4);
        
        x_start = max(1, x);
        x_end = min(width, x + w);
        y_start = max(1, y);
        y_end = min(height, y + h);
        
        for py = y_start:y_end
            for px = x_start:x_end
                if map(py, px) == 255
                    map(py, px) = 180;  % Medium gray for parking
                end
            end
        end
    end
end

function map = add_landing_zones(map, width, height)
    % Add emergency landing zones (keep as free space but mark)
    landing_zones = [
        width*0.1, height*0.5, 60, 60;   % Emergency zone 1
        width*0.9, height*0.5, 60, 60;   % Emergency zone 2
        width*0.5, height*0.1, 80, 80;   % Main landing zone
        width*0.5, height*0.9, 80, 80;   % Secondary landing zone
    ];
    
    for i = 1:size(landing_zones, 1)
        x = round(landing_zones(i, 1));
        y = round(landing_zones(i, 2));
        w = landing_zones(i, 3);
        h = landing_zones(i, 4);
        
        x_start = max(1, x);
        x_end = min(width, x + w);
        y_start = max(1, y);
        y_end = min(height, y + h);
        
        % Clear area for landing
        for py = y_start:y_end
            for px = x_start:x_end
                if map(py, px) > 128  % Don't override buildings/roads
                    map(py, px) = 255;  % Keep as free space
                end
            end
        end
        
        % Add border markers
        border_width = 4;
        map(y_start:y_start+border_width, x_start:x_end) = 150;
        map(y_end-border_width:y_end, x_start:x_end) = 150;
        map(y_start:y_end, x_start:x_start+border_width) = 150;
        map(y_start:y_end, x_end-border_width:x_end) = 150;
    end
end

function save_pgm_file(map_data, filename)
    % Save map as PGM file format for ROS
    fid = fopen(filename, 'w');
    if fid == -1
        error('Could not create file %s', filename);
    end
    
    [height, width] = size(map_data);
    
    % Write PGM header
    fprintf(fid, 'P2\n');
    fprintf(fid, '# Urban City Map for Drone Navigation\n');
    fprintf(fid, '# Generated by MATLAB Urban Map Generator\n');
    fprintf(fid, '%d %d\n', width, height);
    fprintf(fid, '255\n');
    
    % Write pixel data
    for y = 1:height
        for x = 1:width
            fprintf(fid, '%d ', map_data(y, x));
        end
        fprintf(fid, '\n');
    end
    
    fclose(fid);
    fprintf('Saved PGM file: %s\n', filename);
end

function create_visualization(map_data)
    % Create and save visualization
    figure('Name', 'Urban City Map', 'Position', [100, 100, 800, 600]);
    
    % Create colormap
    colormap_custom = [
        0, 0, 0;        % Black - Buildings (value 0)
        0.2, 0.2, 0.2;  % Dark gray - Residential (value 50)
        0.4, 0.4, 0.4;  % Gray - Intersections (value 100)
        0.5, 0.5, 0.5;  % Roads (value 128)
        0.6, 0.6, 0.6;  % Landing zone borders (value 150)
        0.7, 0.7, 0.7;  % Parking (value 180)
        0.8, 0.8, 0.8;  % Parks (value 200)
        1, 1, 1;        % White - Free space (value 255)
    ];
    
    imagesc(map_data);
    colormap(gray);
    axis equal;
    axis tight;
    
    title('Urban City Environment Map for Drone Navigation', 'FontSize', 14, 'FontWeight', 'bold');
    xlabel('X coordinate (pixels)', 'FontSize', 12);
    ylabel('Y coordinate (pixels)', 'FontSize', 12);
    
    % Add legend
    text(100, 100, {'Legend:', 'Black = Buildings', 'Dark Gray = Residential', ...
                    'Gray = Roads', 'Light Gray = Parks', 'White = Free Space'}, ...
         'FontSize', 10, 'BackgroundColor', 'white', 'EdgeColor', 'black');
    
    % Save visualization
    print('urban_city_visualization.png', '-dpng', '-r300');
    fprintf('Saved visualization: urban_city_visualization.png\n');
end 